parser_library/src/workspaces/macro_cache.cpp

```diff
@@ -0,0 +1,202 @@
+    return cached_data.cached_analyzer.get();
+}
+
+bool macro_cache::load_from_cache(const macro_cache_key& key, const analyzing_context& ctx)
+{
+    if (auto cached_analyzer = find_cached_analyzer(key))
+    {
+        if (key.data.proc_kind == processing::processing_kind::MACRO)
+        {
+            lsp::macro_info_ptr info = cached_analyzer->context().lsp_ctx->get_macro_info(key.data.library_member);
+            if (!info)
+                return true; // The file for which the analyzer is cached does not contain definition of macro
+            ctx.hlasm_ctx->add_macro(info->macro_definition);
+            ctx.lsp_ctx->add_macro(info, lsp::text_data_ref_t(macro_file_->get_text()));
+
+            // Add all copy members on which this macro is dependant
+            for (const auto& copy_ptr : info->macro_definition->used_copy_members)
+            {
+                auto copy_file = file_mngr_->find(copy_ptr->definition_location.file);
+                ctx.hlasm_ctx->add_copy_member(copy_ptr);
+                ctx.lsp_ctx->add_copy(copy_ptr, lsp::text_data_ref_t(copy_file->get_text()));
+            }
+        }
+        else if (key.data.proc_kind == processing::processing_kind::COPY)
+        {
+            auto copy_member = cached_analyzer->context().hlasm_ctx->get_copy_member(key.data.library_member);
+            ctx.hlasm_ctx->add_copy_member(copy_member);
+            ctx.lsp_ctx->add_copy(copy_member, lsp::text_data_ref_t(macro_file_->get_text()));
+        }
+
+        return true;
+    }
+    return false;
+}
+
+version_stamp macro_cache::get_copy_member_versions(context::macro_def_ptr macro) const
+{
+    version_stamp result;
+
+    for (const auto& copy_ptr : macro->used_copy_members)
+    {
+        auto file = file_mngr_->find(copy_ptr->definition_location.file);
+        if (!file)
+            throw std::runtime_error("Dependencies of a macro must be open right after parsing the macro.");
+        result.try_emplace(file->get_file_name(), file->get_version());
+    }
+    return result;
+}
+
+void macro_cache::save_analyzer(const macro_cache_key& key, std::unique_ptr<analyzer> analyzer)
+{
+    auto& cache_data = cache_[key];
+    if (key.data.proc_kind == processing::processing_kind::MACRO)
+    {
+        // Add stamps for all macro dependencies
+        auto parsed_macro = analyzer->context().hlasm_ctx->get_macro_definition(key.data.library_member);
+        if (parsed_macro)
+            cache_data.stamps = get_copy_member_versions(std::move(parsed_macro));
+        else
+            cache_data.stamps.clear();
+    }
+    else // Copy members do not have additional dependencies
+        cache_data.stamps.clear();
+
+    cache_data.stamps.try_emplace(macro_file_->get_file_name(), macro_file_->get_version());
+    cache_data.cached_analyzer = std::move(analyzer);
+}
+
+void macro_cache::erase_cache_of_opencode(const std::string opencode_file_name)
+{
+    auto it = cache_.begin();
```

Comment:
any concurrency concerns?

Reply:
the function is currently called from the workspace, which does not have any concurrent usage, as far as I know